src/spec.rs

```diff
@@ -52,69 +122,132 @@ impl CentralDirectoryEnd {
     }
 
     pub fn find_and_parse<T: Read + Seek>(reader: &mut T) -> ZipResult<(CentralDirectoryEnd, u64)> {
-        const HEADER_SIZE: u64 = 22;
-        const BYTES_BETWEEN_MAGIC_AND_COMMENT_SIZE: u64 = HEADER_SIZE - 6;
         let file_length = reader.seek(io::SeekFrom::End(0))?;
 
-        let search_upper_bound = file_length.saturating_sub(HEADER_SIZE + u16::MAX as u64);
-
-        if file_length < HEADER_SIZE {
+        if file_length < CDEBlock::BLOCK_LEN as u64 {
             return Err(ZipError::InvalidArchive("Invalid zip header"));
         }
 
-        let mut pos = file_length - HEADER_SIZE;
-        while pos >= search_upper_bound {
-            reader.seek(io::SeekFrom::Start(pos))?;
-            if reader.read_u32_le()? == CENTRAL_DIRECTORY_END_SIGNATURE {
-                reader.seek(io::SeekFrom::Current(
-                    BYTES_BETWEEN_MAGIC_AND_COMMENT_SIZE as i64,
-                ))?;
-                let cde_start_pos = reader.seek(io::SeekFrom::Start(pos))?;
-                if let Ok(end_header) = CentralDirectoryEnd::parse(reader) {
-                    return Ok((end_header, cde_start_pos));
+        let search_upper_bound = 0;
+
+        const END_WINDOW_SIZE: usize = 512;
+
+        let sig_bytes = CENTRAL_DIRECTORY_END_SIGNATURE.to_le_bytes();
+        let finder = FinderRev::new(&sig_bytes);
+
+        let mut window_start: u64 = file_length.saturating_sub(END_WINDOW_SIZE as u64);
+        let mut window = [0u8; END_WINDOW_SIZE];
+        let mut has_tried_zero: bool = false;
+        while window_start >= search_upper_bound {
+            if has_tried_zero {
```

Comment:
You can eliminate this by moving the `if window_start == search_upper_bound` check to the bottom of the loop.